### ISSUE TITLE
`CubedSphereData` and `CubedSphereBoundaryConditions` abstractions

### ISSUE THREAD

**Issue by ali-ramadhan** [CliMA/Oceananigans.jl#1583]
These would be useful abstractions so that cubed sphere models can reuse all the existing field types in `Oceananigans.Fields` without introducing any new messy field types.

**Comment by glwagner** [comment]
Can you summarize the capabilities that the abstraction will need? I understand it will contain references to the local parts of the underlying arrays for all 6 faces... ?

**Comment by ali-ramadhan** [comment]
It'll probably need to store N arrays under the hood where N is the number of faces.

Right now we use 6 faces but if you're doing cubed sphere + MPI you might end up doing 1 < N < 6 faces on each rank.

Otherwise I think the abstraction should enable users to use the cubed sphere grid like they would any other grid. So things like `set!` should just work.

Right now it's more about making sure the backend all works with `CubedSphereData` then a nice user interface can be built on top of it.

There are some decisions to make like should `interior(::Field{..., <:CubedSphereData})` return a 4D array? But we can decide on those as they come up.

It'll be nice if the output writers knew how to write `CubedSphereData` to disk, although this can come later and many users may want the fields interpolated onto a regular lat-lon grid which is a different discussion.

**Comment by glwagner** [comment]
> There are some decisions to make like should interior(::Field{..., <:CubedSphereData}) return a 4D array?

To be sure, the question is whether it should return a _view_ to a 4D array (not a 4D array itself). `interior` currently returns a view.

I'm pretty confused about how this would work with MPI.

Another question is what to return from `parent(field::Field{X, Y, Z, A, <:CubedSphereGrid})`. For example, we have some code that uses things like

```julia
parent(a) .= parent(b)
```

Can we get broadcasting to work with whatever `parent` returns? (An alternative strategy is to get broadcasting working generically so we just write `a .= b` and forget about `parent`).
